pilot/pkg/model/virtualservice.go

```diff
@@ -0,0 +1,348 @@
+	}
+	for k, v := range root.SourceLabels {
+		out.SourceLabels[k] = v
+	}
+	for k, v := range delegate.SourceLabels {
+		out.SourceLabels[k] = v
+	}
+	if len(out.Gateways) == 0 {
+		out.Gateways = root.Gateways
+	}
+	if out.SourceNamespace == "" {
+		out.SourceNamespace = root.SourceNamespace
+	}
+	if out.Port == 0 {
+		out.Port = root.Port
+	}
+	return &out
+}
+
+func hasConflict(root, leaf *networking.HTTPMatchRequest) bool {
+	roots := []*networking.StringMatch{root.Uri, root.Scheme, root.Method, root.Authority}
+	leaves := []*networking.StringMatch{leaf.Uri, leaf.Scheme, leaf.Method, leaf.Authority}
+	for i := range roots {
+		if stringMatchConflict(roots[i], leaves[i]) {
+			return true
+		}
+	}
+	// header conflicts
+	for key, leafHeader := range leaf.Headers {
+		if stringMatchConflict(root.Headers[key], leafHeader) {
+			return true
+		}
+	}
+
+	// without headers
+	for key, leafValue := range leaf.WithoutHeaders {
+		if stringMatchConflict(root.Headers[key], leafValue) {
+			return true
+		}
+	}
+
+	// query params conflict
+	for key, value := range leaf.QueryParams {
+		if stringMatchConflict(root.QueryParams[key], value) {
+			return true
+		}
+	}
+
+	// source labels
+	for key, leafValue := range leaf.SourceLabels {
+		if rootValue, ok := root.SourceLabels[key]; ok && rootValue != leafValue {
+			return true
+		}
+	}
+
+	if len(leaf.Gateways) > 0 {
+		leafGws := sets.NewSet(leaf.Gateways...)
+		for _, gw := range root.Gateways {
+			if leafGws.Contains(gw) {
+				return true
+			}
+		}
+	}
+
+	if root.IgnoreUriCase != leaf.IgnoreUriCase {
+		return true
+	}
+	if root.SourceNamespace != "" && leaf.SourceNamespace != "" && root.SourceNamespace != leaf.SourceNamespace {
+		return true
+	}
+	if root.Port > 0 && leaf.Port > 0 && root.Port != leaf.Port {
+		return true
+	}
+
+	return false
+}
+
+func stringMatchConflict(root, leaf *networking.StringMatch) bool {
+	// no conflict when root or leaf is not specified
+	if root == nil || leaf == nil {
+		return false
+	}
+	// regex match is not allowed
+	if root.GetRegex() != "" || leaf.GetRegex() != "" {
+		return true
+	}
+	// root is exact match
+	if exact := root.GetExact(); exact != "" {
+		// leaf is prefix match, conflict
+		if leaf.GetPrefix() != "" {
+			return true
+		}
+		// both exact, but not equal
+		if leaf.GetExact() != exact {
+			return true
+		}
+		return false
+	}
+	// root is prefix match
+	if prefix := root.GetPrefix(); prefix != "" {
+		// leaf is prefix match
+		if leaf.GetPrefix() != "" {
+			// leaf(`/a`) is not subset of root(`/a/b`)
+			if !strings.HasPrefix(leaf.GetPrefix(), prefix) {
+				return true
+			}
+			return false
+		}
+		// leaf is exact match
+		if leaf.GetExact() != "" {
+			// leaf(`/a`) is not subset of root(`/a/b`)
+			if !strings.HasPrefix(leaf.GetExact(), prefix) {
+				return true
+			}
+			return false
+		}
+	}
+
+	return true
+}
+
+func mergeHeaders(root *networking.Headers, leaf *networking.Headers) *networking.Headers {
```

Comment:
This is misleading? Its not merging

Reply:
I need to make it align with other fields

Reply:
Done